### PR TITLE
Download and extract native tools to temporary locations.

### PR DESCRIPTION
**Pull request by AustinWise**
Consider the following scenario:
1. Build a repo using native tools, such as dotnet/wpf.
1. While a zip file is downloading (for example in wpf, strawberry-perl), the build is aborted by pressing Ctrl+C.
1. The build is run again.

### Expected behavior
The zip file is downloaded from scratch.

### Actual behavior
The build process attempts to extract the partially downloaded zip file. This fails of course. `git clean -dxf` does not fix the situtation, as these zip files are stored under `%USERPROFILE%` somewhere.

### This PR's changes
The zip file is first downloaded to a temporary path, then moved to its final destination. Additionally when extracting the zip file, it is first extracted to a temp folder before moving to the final destination.

### CODE REVIEW

**Comment by jonfortescue** [comment]
also adding @mjanecke and @alexperovich for powershell review if you have time.

File: eng/common/native/CommonLibrary.psm1 [hunk 1]
```diff
@@ -328,16 +331,21 @@ function Expand-Zip {
...
       }
     }
-    if (-Not (Test-Path $OutputDirectory)) {
-      New-Item -path $OutputDirectory -Force -itemType "Directory" | Out-Null
+
+    $TempOutputDirectory = Join-Path "$(Split-Path -Parent $OutputDirectory)" "$(Split-Path -Leaf $OutputDirectory).tmp"
+    if (Test-Path $TempOutputDirectory) {
+      Remove-Item $TempOutputDirectory -Force -Recurse
     }
+    New-Item -path $TempOutputDirectory -Force -itemType "Directory" | Out-Null
```
Comment (chcosta) [inline eng/common/native/CommonLibrary.psm1]: nit: convention in this file would be `-Path` and `-ItemType`

File: eng/common/native/CommonLibrary.psm1 [hunk 2]
```diff
@@ -116,17 +116,20 @@ function Get-File {
     New-Item -path $DownloadDirectory -force -itemType "Directory" | Out-Null
   }
 
+  $TempPath = "$Path.tmp"
   if (Test-Path -IsValid -Path $Uri) {
     Write-Verbose "'$Uri' is a file path, copying file to '$Path'"
-    Copy-Item -Path $Uri -Destination $Path
+    Copy-Item -Path $Uri -Destination $TempPath
```
Comment (chcosta) [inline eng/common/native/CommonLibrary.psm1]: This seems unnecessary

File: eng/common/native/CommonLibrary.psm1 [hunk 3]
```diff
@@ -116,17 +116,20 @@ function Get-File {
     New-Item -path $DownloadDirectory -force -itemType "Directory" | Out-Null
   }
 
+  $TempPath = "$Path.tmp"
   if (Test-Path -IsValid -Path $Uri) {
     Write-Verbose "'$Uri' is a file path, copying file to '$Path'"
```
Comment (mjanecke) [inline eng/common/native/CommonLibrary.psm1]: Do you want to rename this $Path to also $TempPath?

**Comment by jonfortescue** [comment]
@AustinWise this looks good to merge once feedback is addressed.

File: eng/common/native/CommonLibrary.psm1 (hunk 2 above)
Comment (AustinWise) [reply eng/common/native/CommonLibrary.psm1]: A large file copy could be interrupted partway through and suffer from the same incomplete file problem as a download.

File: eng/common/native/CommonLibrary.psm1 (hunk 3 above)
Comment (AustinWise) [reply eng/common/native/CommonLibrary.psm1]: I pushed an updates that make the these verbose logs mention the temporary files.

File: eng/common/native/CommonLibrary.psm1 (hunk 1 above)
Comment (AustinWise) [reply eng/common/native/CommonLibrary.psm1]: I pushed an update that fixes this.

**Comment by AustinWise** [comment]
@jonfortescue I pushed an update that addresses some of the feedback and left responses to @chcosta's and @mjanecke's feedback.

**Comment by vatsan-madhavan** [comment]
WPF builds are seeing occasional problems due to failures during extraction - see https://github.com/dotnet/core-eng/issues/7072. 

This PR would be a great addition to the native-tools download & extraction process! 

/cc @mattgal, @rladuca

**Comment by MattGal** [comment]
@vatsan-madhavan this is nice to do, but unless a distinct different temp path were used every time the problem we hit in dotnet/core-eng#7072 would still be there, specifically if the thing downloaded ever is HTML labelled zip, the caching behavior (which is super nice to have) would remain and cause trouble.

**Comment by jonfortescue** [comment]
@chcosta is this good to merge?

**Comment by jonfortescue** [comment]
@chcosta @AustinWise should this be merged now? Looks like all is passing.

**Comment by AustinWise** [comment]
@jonfortescue I'm not sure why these are failing. So far all of them seem to be in the form of:

```
Test HelixReporterTests.failure1 was expected to fail but did not fail.
```

I tried running the tests again locally (rebased onto master), but then I get a completely different error for `src/Microsoft.DotNet.SignTool.Tests/Microsoft.DotNet.SignTool.Tests.csproj`:

```
It was not possible to find any compatible framework version
The framework 'Microsoft.NETCore.App', version '2.1.0' was not found.
  - The following frameworks were found:
      3.1.1 at [/tank/externsrc/dotnet/arcade/.dotnet/shared/Microsoft.NETCore.App]

You can resolve the problem by installing the specified framework and/or SDK.
```

This looks like some sort of test authoring or build script problem.

Let me know if I should look into this deeper.

**Comment by JohnTortugo** [comment]
From a quick look the failures seem unrelated to your change. Can you try rebasing your branch?

**Comment by AustinWise** [comment]
@JohnTortugo I rebased and squashed my changes. It looks like CI passed this time.

**Pull request merged**